Support `SparseObservable` to `SparsePauliOp` conversions

Cryoris opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

Add `SparseObservable.to_sparse_list` to generate a sparse Pauli list representation, which is used to implement `SparsePauliOp.from_sparse_observable`. This method will also allow using `SparseObservable`s in a `PauliEvolutionGate` (follow up PR).

### Details and comments

In as many places as I could, I added a warning that this conversion will blow up if one attempts to convert a large number of projectors to a Pauli representation.

Cryoris committed:
SparseObservable -> SparsePauliOp

and SparseObservable.to_sparse_list, which can also be used for the PauliEvolutionGate compat

Cryoris committed:
reno

Cryoris committed:
allow observable alphabet as well

Cryoris committed:
Merge branch 'main' into observable-interconnect

jakelishman commented on crates/accelerate/src/sparse_observable.rs:
```diff
@@ -641,6 +660,46 @@ impl SparseObservable {
+    /// This representation is highly inefficient for projectors. For example, a term with
+    /// :math:`n` projectors :math:`|+\rangle\langle +|` will use :math:`2^n` Pauli terms.
+    pub fn to_paulis(&self) -> impl Iterator<Item = (Vec<BitTerm>, Vec<u32>, Complex64)> + '_ {
+        self.iter().flat_map(|view| {
```
How would you feel about having the `to_paulis` method return another `SparseObservable`, just with the `BitTerm`s rewritten in the way you have now?  We could expose it to Python in that form.

I'm just thinking: this could simplify the `to_sparse_list` method, because then we wouldn't need the `only_paulis` option at all: it's the composition `obs.to_paulis().to_sparse_list()`.

The `to_paulis()` method would also have a secondary use: expensive as it is, `to_paulis().simplify()` produces a unique canonical form for any given `SparseObservable`; everything representing the same observable has the same form under that method, because the Paulis are a precise basis, rather than overcomplete.

Cryoris replied:
Yeah I also considered that, the downside is that we need to take a detour via `SparsePauliOp` if we want the sparse list for the Pauli evolution. So we'd always have to go via the "dense" format where we store all identities explicitly, which would be nice to avoid 🙂 

The alternative would be to handle projectors in the Pauli evolution directly, but from our offline discussion yesterday I'm not sure if we're overcomplicating things then 🤔

jakelishman replied:
I mean having `to_paulis` return `SparseObservable`, not `SparsePauliOp` - still qubit-sparse, but with a promise that we only use alphabet letters corresponding to Pauli ops?

jakelishman replied:
Not sure if you misunderstood me before, or if I misunderstood you.  Returning `SparseObservable` from this method is actually slightly better in terms of memory use than the collection into a list/vec of this output, and I'm not super concerned about the intermediate memory usage of the `SparseObservable` object here, because the usage of the resulting `SparsePauliOp` is already going to be about the same (depends on how dense the terms are).

Cryoris replied:
Ah! I misunderstood you! Oh yeah having `to_paulis` return a `SparseObservable` (not `SparsePauliOp`) would be an elegant solution indeed. I'll update that 🙂

Cryoris replied:
Changed in 358575b, seems much nicer now 🙂

jakelishman commented on qiskit/quantum_info/operators/symplectic/sparse_pauli_op.py:
```diff
@@ -929,6 +931,45 @@ def from_sparse_list(
+                    f"lead to at least {2 ** num_projectors} terms.",
+                    stacklevel=2,
+                    category=RuntimeWarning,
+                )
```
The actual runtime warnings might be a shade too far, maybe - I think the `.. warning::` note you've got in the documentation is probably enough.

Cryoris replied:
Sounds good -- I just really wanted to make sure people don't do this 😛

Cryoris replied:
Changed in 358575b 👍🏻

jakelishman commented on test/python/quantum_info/operators/symplectic/test_sparse_pauli_op.py:
```diff
@@ -361,6 +367,36 @@ def test_to_list_parameters(self):
+
+        obs = SparseObservable("XrZ")
+        expected = SparsePauliOp(["XIZ", "XYZ"], coeffs=[0.5, -0.5])
+        self.assertEqual(expected, SparsePauliOp.from_sparse_observable(obs))
```
This test is simple enough to be ok, but just note that technically it's potentially coupling to an internal detail of `SparsePauliOp.from_sparse_observable`: `SparsePauliOp(["XYZ", "XIZ"], coeffs=[-0.5, 0.5]` would also be a correct answer, but the test would reject it.

We could say:
- `SparseObservable.to_paulis()` will be in fully sorted canonical order.  This could be quite a lot of additional overhead.
- `SparseObservable.to_paulis()` will always expand each individual projector `BitTerm` in canonical order (sorts to IXYZ, etc), in which case the order is completely defined, but not fully canonical.
- there is no defined order from `SparseObservable.to_paulis()`, in which case this test should include a canonicalisation step after `SparsePauliOp.from_sparse_observable(obs)` to isolate the test (and the others in this file) from changes in the order.

We ought to pick one of those - do you have a preference?

Cryoris replied:
I would have a preference for the last option -- if there's no clear practical benefit in guaranteeing a format then I'd rather not commit to one 🙂

jakelishman replied:
Fine by me - in that case let's update the tests to decouple them.

Cryoris replied:
done in 358575b

jakelishman commented on test/python/quantum_info/test_sparse_observable.py:
```diff
@@ -2008,3 +2008,71 @@ def test_term_pauli_base(self):
+        sparse_list = obs.to_sparse_list()
+        with self.subTest(msg="identity"):
+            self.assertEqual(1, len(sparse_list))
+            self.assertEqual("", sparse_list[0][0])
```
Can we write these equality tests (etc) using the actual expected lists - that'll test length and full correctness in a single step, and give better error messages.

Cryoris replied:
Done in 358575b

jakelishman commented on test/python/quantum_info/test_sparse_observable.py:
```diff
@@ -2008,3 +2008,71 @@ def test_term_pauli_base(self):
+        with self.subTest(msg="IXYZ"):
+            self.assertEqual(1, len(sparse_list))
+            self.assertEqual("ZYX", sparse_list[0][0])
+            self.assertListEqual([0, 1, 2], sparse_list[0][1])
```
You typically don't need/want to call `assertListEqual` (just `assertEqual`).  If `sparse_list[0][1]` is some list-like type and not necessarily _precisely_ a list, then `self.assertEqual([0, 1, 2], list(x))` is fine.

Cryoris replied:
Done in 358575b

jakelishman commented on test/python/quantum_info/operators/symplectic/test_sparse_pauli_op.py:
```diff
@@ -361,6 +367,36 @@ def test_to_list_parameters(self):
 
+    def test_from_sparse_observable(self):
+        """Test from a SparseObservable."""
+        obs = SparseObservable.zero(10)
```
Can we also add a test of the zero-qubit zero operator and zero-qubit identity as edge cases?

Cryoris replied:
Added in 358575b

jakelishman reviewed (commented):
Thanks for doing this - one main question about the exposed interfaces, and a couple of minor bits on testing.

Cryoris committed:
Jake's review comments

- to_paulis -> SparseObservable
- test more edge cases
- don't guarantee any ordering

Cryoris committed:
Merge branch 'observable-interconnect' of github.com:Cryoris/qiskit-terra into observable-interconnect

Cryoris committed:
update reno

Cryoris committed:
fix ruff

... which was somehow not captured by my pre commit hook

jakelishman commented on crates/accelerate/src/sparse_observable.rs:
```diff
@@ -641,6 +660,52 @@ impl SparseObservable {
+        }
+
+        SparseObservable::new(self.num_qubits, coeffs, paulis, indices, boundaries)
+    }
```
This function should never return a `CoherenceError` - the input `SparseObservable` is guaranteed valid, and all valid `SparseObservable`s have a valid representation in the pure-Pauli-operator basis.  We can construct the output `SparseObservable` directly using the record-struct constructor (`SparseObservable {num_qubits: self.num_qubits, ..}`) or using `SparseObservable::new_unchecked` with a mild preference on my side for the record-struct constructor.

jakelishman commented on crates/accelerate/src/sparse_observable.rs:
```diff
@@ -184,6 +185,24 @@ impl BitTerm {
+        BitTerm::Zero => vec![(true, None), (true, Some(BitTerm::Z))],
+        BitTerm::One => vec![(true, None), (false, Some(BitTerm::Z))],
+    }
 }
```
Rather unimportant, but these could all return `&'static [(bool, Option<BitTerm>)]`, rather than eagerly allocating a `Vec`.

jakelishman commented on crates/accelerate/src/sparse_observable.rs:
```diff
@@ -2337,6 +2402,63 @@ impl PySparseObservable {
+    ///
+    /// This can be seen as counter-operation of :meth:`.SparseObservable.from_sparse_list`, however
+    /// the order of terms is not guaranteed to be the same at after a roundtrip to a sparse
+    /// list and back. Use :meth:`SparseObservable.simplify` to obtain a canonical representation.
```
I think this note about `SparseObservable.simplify` actually belongs in the `to_paulis` method above - the recipe for a completely canonical representation of a `SparseObservable` is `obs.to_paulis().simplify()`, nothing to do with `to_sparse_list`.

Cryoris replied:
Good point, I'll add it to the `as_paulis` method. For `to_sparse_list` I also thought we didn't want to guarantee an order, so I left the comment that doing the roundtrip `SparseObservable -> sparse list -> SparseObservable` might not maintain the order of terms.

jakelishman commented on crates/accelerate/src/sparse_observable.rs:
```diff
@@ -2337,6 +2402,63 @@ impl PySparseObservable {
+            .collect::<PyResult<Vec<_>>>()?;
+
+        let out = PyList::new(py, sparse_list)?;
+        Ok(out.unbind())
```
Two points here, both pretty minor:

1. why deconstruct the `SparseView` _before_ `to_py_tuple` and not just accept that in the closure?
2. it'd probably be a bit more efficient to construct a `let out = PyList::empty(py)?;` and do `for term in inner.iter(): out.append(to_py_tuple(term)?)?;` than to collect into a `Vec` and then convert that into a list

Cryoris replied:
1. that's legacy from when `to_paulis` returned the bits/indices/coeffs separately
2. good idea, I was looking for a way to avoid this!

Cryoris replied:
Done in 984a75c54

jakelishman commented on qiskit/quantum_info/operators/symplectic/sparse_pauli_op.py:
```diff
@@ -929,6 +930,27 @@ def from_sparse_list(
         paulis = PauliList(labels)
         return SparsePauliOp(paulis, coeffs, copy=False)
 
+    @staticmethod
+    def from_sparse_observable(obs: SparseObservable) -> SparsePauliOp:
```
Do you want to add a variant to the `__init__` method that handles this case too?  `quantum_info` types (for better or worse) check everything under the sun in their default constructors.

Cryoris replied:
I think we said we don't want to add this because it would give a very easy path to a potentially very expensive representation -- so it's after a "you-need-an-explicit-constructor"-wall 😄

jakelishman replied:
Ok sure, that's valid too.  I forgot we'd said that haha

jakelishman commented on crates/accelerate/src/sparse_observable.rs:
```diff
@@ -2337,6 +2402,63 @@ impl PySparseObservable {
+            }
+            let py_string = PyString::new(py, &pauli_string).unbind();
+            let py_indices = PyList::new(py, indices)?.unbind();
+            let py_coeff = coeff.into_py_any(py)?;
```
What do you think about reversing the bit terms and the indices on output, so the Pauli string comes out in "reading" order?

Minor, but the necessary capacity of `pauli_string` is known immediately; all the `py_label` return values are (necessarily) a single ASCII letter, so `pauli_string` needs a capacity of `bits.len()`, and we could use `String::with_capacity` on it.

Cryoris replied:
Sounds good, that'd be a bit more intuitive 🙂 Also that'll verify that the tests are indeed order-independent 😉 

Good call on the `with_capacity` 👍🏻

Cryoris replied:
Done in 984a75c54

jakelishman commented on crates/accelerate/src/sparse_observable.rs:
```diff
@@ -2337,6 +2402,63 @@ impl PySparseObservable {
+    ///     >>> obs = SparseObservable("+")
+    ///     >>> obs.to_paulis()
+    ///     <SparseObservable with 2 terms on 1 qubit: (0.5+0j)() + (0.5+0j)(X_0)>
+    fn to_paulis(&self) -> PyResult<Self> {
```
Very minor: the name `to_paulis` in the context of Qiskit makes me think it might come out related to the `Pauli` class.  _Perhaps_ `as_paulis` makes it clearer from reading that it remains a `SparseObservable`?

Cryoris replied:
Done in 984a75c54

jakelishman reviewed (commented):
Thanks for doing this - it'll be a big help for tests, and it's good to add the conversion, even if it's expensive in some cases.

We need to add the new methods into sensible places into the class-level documentation - they're not linked at the moment.

Cryoris committed:
round 2

- use &'static over Vec
- reverse bit/index order
- rename as_paulis
- direct construciton w/o CoherenceError

Cryoris committed:
fix tests and improve docs

jakelishman committed:
Update documentation

jakelishman reviewed (approved):
Thanks for this.  I tweaked the docs slightly to add a couple more examples, and modify the cross-links a little bit, but other than that it looks good to go to me.
